apps/evm/lib/block/header.ex

```diff
@@ -183,11 +183,11 @@ defmodule Block.Header do
       transactions_root: transactions_root,
       receipts_root: receipts_root,
       logs_bloom: logs_bloom,
-      difficulty: :binary.decode_unsigned(difficulty),
```

Comment:
since this is getting called a lot, are you sure it makes sense to wrap it in a if statement wrapper?

Reply:
If statement? Do you mean the guard-based  check to verify its type?

The problem is this:

```elixir
[5, <<5>>] |> ExRLP.encode |> ExRLP.decode
[<<5>>, <<5>>]
```

ExRLP can't determine if a value was a binary or a decimal. So, we could always decode, but then it breaks code that does this:

```
MyPacket.serialize() # serialize returns an RLP-encodable structure
[5, 5, 6]
MyPacket.deserialize([5, 5, 6])
** (ArgumentError) argument error
    (stdlib) :binary.decode_unsigned(5)
```

We'd expect serialize and deserialize to work.

So maybe we could enforce that serialize always returns binaries, but that's not technically a requirement of the RLP specification... and would tend to make our lives harder.

So instead.... we just add this check and decode only if it's not already an integer.

Reply:
Ahaaa. I didn't know it was breaking! Looking at the old code I thought we *knew* it's a binary and that we're now just double checking it is :)